sql/core/src/main/scala/org/apache/spark/sql/execution/python/ExtractPythonUDFs.scala

```diff
@@ -169,24 +169,64 @@ object ExtractPythonUDFs extends Rule[LogicalPlan] with Logging {
     e.exists(PythonUDF.isScalarPythonUDF)
   }
 
+  /**
+   * Return true if we should extract the current expression, including all of its current
+   * children (including UDF expression, and all others), to a logical node.
+   * The children of the expression can be udf expressions, this would be `chaining`.
+   * If child udf expressions were already extracted before, then this will just extract
```

Comment:
I think this one is a little bit hard to follow. Would be much easier if we can put some ascii blocks and explanation.